### ISSUE TITLE
feat: `defaultValue` should allow an async function

### ISSUE THREAD

**Issue by diosney** [sequelize/sequelize#4569]
This will allow for checking on instance values and other models instances values.

Right now the way I accomplish this is with a hook.

**Comment by mickhansen** [comment]
This will likely not be implemented for a good while, if at all. It's solved pretty well by hooks and it would require a less than trivial amount of refactoring.

**Comment by diosney** [comment]
@mickhansen No problem. Can you let this issue open as a remainder until it is done? Thanks!

**tldr;** Right now I solved the issue as you said, with hooks, but even hooks has some caveats for setting the initial value.

(I explained all the things below just to let you know why a async `defaultValue`  could be awesome and to provide you more information if I missed something important, like an easier solution)

In my case:

* I've two models A and B
* Model A has an attribute `attributeA` with some value already set.
* Model B has an attribute `attributeB` with `allowNull: false`, this is to ensure that always a value is set, even if its value is async taken from another instance attribute.
* On model B I want that if some attribute is not set, then take the value from `instanceA.attributeA`.

Now, the most logical hook where a default value should be implemented is `beforeCreate`, so the other sister hooks `beforeDestroy`, `beforeUpdate` dont' execute the hook function, but this conflicts with `ModelB.attributeB` `allowNull: false`, since `validate` is run way before `beforeCreate`, so it will trigger an error when `attributeB` isn't set.

Now, to overcome this I moved the functionality from `beforeCreate` to `beforeValidate`, but then this hook will execute for all operations (create, update, delete) instead of only when creating new instances.

I know, I added a check to warding against this in the `beforeValidate` hook, something like `if value is not set, then...`, but even this conditional will execute for every operation hook, which can be avoided if:

* `defaultValue` accepts an async function, or
* the hook order was something like:

```
(1)
  beforeBulkCreate(instances, options, fn)
  beforeBulkDestroy(instances, options, fn)
  beforeBulkUpdate(instances, options, fn)
(2)
  beforeCreate(instance, options, fn)
  beforeDestroy(instance, options, fn)
  beforeUpdate(instance, options, fn)
  beforeValidate(instance, options, fn)
(-)
  validate
(3)
  afterValidate(instance, options, fn)
(-)
  create
  destroy
  update
(4)
  afterCreate(instance, options, fn)
  afterDestroy(instance, options, fn)
  afterUpdate(instance, options, fn)
(5)
  afterBulkCreate(instances, options, fn)
  afterBulkDestroy(instances, options, fn)
  afterBulkUpdate(instances, options, fn)
```

Over the two options I prefer the first is the better :) since the hook order change can be more cumbersome and API-breaking.

Thanks

**Comment by AdityaAnand1** [comment]
Just ran into this, I agree that using hooks is a possible workaround but what I'm really doing in the beforeCreate is just adding the default value which is what the defaultValue is meant to be for. If it's not too much work, this would be an excellent addition 😄

**Comment by frlinw** [comment]
3 years later, I agree, if defaultValue accept function, it should accept async function

**Comment by bradennapier** [comment]
RIDICULOUS that everything sequelize is Promise-based then it cant support this.  What if you want to capture a value from the database ?? T he shortcomings of this entire library cause some major problems in terms of performance when defining models.
